### PR TITLE
[WIP] Structs for Gateway Events

### PR DESCRIPTION
**Pull request by skippi**
This introduces structs for compile time guarantees on Gateway events.

Very WIP for now. Will be asking for code review periodically. Might even break this up into a couple PRs depending on the scale.

### CODE REVIEW

File: lib/nostrum/gateway/event.ex [hunk 1]
```diff
@@ -0,0 +1,22 @@
+defmodule Nostrum.Gateway.Event do
+  alias Nostrum.Gateway.{HelloEvent, ReadyEvent}
+
+  @type t :: ReadyEvent.t()
+
+  @spec new(Enum.t()) :: {:ok, struct} | :error
+  def new(enum) do
+    string_map =
+      enum
+      |> Enum.map(fn {k, v} -> {to_string(k), v} end)
```
Comment (Kraigie) [inline lib/nostrum/gateway/event.ex]: Why are we converting to strings here and elsewhere throughout these modules?

File: lib/nostrum/gateway/hello_event.ex [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    :_trace
+  ]
+
+  defmodule Field do
+    @moduledoc false
+
+    @spec new({term, term}) :: Keyword.t()
+    def new(pair)
+    def new({k, v}) when not is_binary(k), do: new({to_string(k), v})
+    def new({"heartbeat_interval", value}), do: [heartbeat_interval: value]
```
Comment (Kraigie) [inline lib/nostrum/gateway/hello_event.ex]: What's the reasoning here behind converting these to a kwl? Couldn't the same be done with a map?

File: lib/nostrum/gateway/ready_event.ex [hunk 3]
```diff
@@ -0,0 +1,54 @@
+defmodule Nostrum.Gateway.ReadyEvent do
+  alias Nostrum.Util
+  alias Nostrum.Struct.Channel
```
Comment (Kraigie) [inline lib/nostrum/gateway/ready_event.ex]: 👀

File: lib/nostrum/gateway/event.ex [hunk 4]
```diff
@@ -0,0 +1,22 @@
+defmodule Nostrum.Gateway.Event do
+  alias Nostrum.Gateway.{HelloEvent, ReadyEvent}
+
+  @type t :: ReadyEvent.t()
+
+  @spec new(Enum.t()) :: {:ok, struct} | :error
```
Comment (Kraigie) [inline lib/nostrum/gateway/event.ex]: Is `Enum.t` correct here? We don't support users passing in a list.

File: lib/nostrum/gateway/event.ex [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+    type = to_string(string_map["t"])
+
+    build_event(type, string_map)
+  end
+
+  defp build_event(type, enum)
+  defp build_event("HELLO", enum), do: {:ok, HelloEvent.new(enum)}
+  defp build_event("READY", enum), do: {:ok, ReadyEvent.new(enum)}
+  defp build_event(_, _), do: :error
```
Comment (Kraigie) [inline lib/nostrum/gateway/event.ex]: I'm not sure how we'll end up using these functions, but we might want to be more descriptive here. Maybe a logger call or something, maybe just pass whatever would be here through.

File: lib/nostrum/gateway/event.ex (hunk 1 above)
Comment (skippi) [reply lib/nostrum/gateway/event.ex]: Did a benchmark just now on this. String comparison is too slow, so I'm going to change this to compare with atoms.

**Comment by jchristgit** [comment]
Hi @skippi, how are you doing? How is life treating you? I hope you're doing well with the COVID-19 chaos creeping around. Can we help you out with this pull request? Do you want an Emotional Support Cat?

Thanks,
Johannes

**Comment by jchristgit** [comment]
Closing per our discussion on Discord.

**Pull request closed without merging**